src/zeroconf/_services/browser.py

```diff
@@ -250,43 +316,115 @@ def _generate_first_next_time(self, now: float_) -> None:
-        """Reschedule the query for a type to happen sooner."""
-        if next_time >= self._next_time[type_]:
-            return False
-        self._next_time[type_] = next_time
-        return True
-
-    def _force_reschedule_type(self, type_: str_, next_time: float_) -> None:
-        """Force a reschedule of a type."""
-        self._next_time[type_] = next_time
-
-    def process_ready_types(self, now: float_) -> List[str]:
+        start_delay = millis_to_seconds(random.randint(*self._first_random_delay_interval))
+        self._loop = loop
+        self._next_run = loop.call_later(start_delay, self._process_startup_queries)
+
+    def stop(self) -> None:
+        """Stop the scheduler."""
+        if self._next_run is not None:
+            self._next_run.cancel()
+            self._next_run = None
+        self._next_scheduled_for_alias.clear()
+        self._query_heap.clear()
+
+    def schedule(self, pointer: DNSPointer) -> None:
+        """Schedule a query for a pointer."""
+        self._schedule(pointer, pointer.get_expiration_time(_EXPIRE_REFRESH_TIME_PERCENT))
+
+    def _schedule(self, pointer: DNSPointer, expire_time: float_) -> None:
+        """Schedule a query for a pointer."""
+        scheduled_query = _ScheduledPTRQuery(pointer.alias, pointer.name, expire_time)
+        self._next_scheduled_for_alias[pointer.alias] = scheduled_query
+        heappush(self._query_heap, scheduled_query)
+
+    def cancel(self, pointer: DNSPointer) -> None:
+        """Cancel a query for a pointer."""
+        scheduled = self._next_scheduled_for_alias.pop(pointer.alias, None)
+        if scheduled:
+            scheduled.cancelled = True
+
+    def reschedule(self, pointer: DNSPointer) -> None:
+        """Reschedule a query for a pointer."""
+        current = self._next_scheduled_for_alias.get(pointer.alias)
+        expire_time = pointer.get_expiration_time(_EXPIRE_REFRESH_TIME_PERCENT)
+        if current is not None:
+            # If the expire time is within self._min_time_between_queries_millis
+            # of the current scheduled time avoid churn by not rescheduling
+            if (
+                -self._min_time_between_queries_millis
+                <= expire_time - current.when_millis
+                <= self._min_time_between_queries_millis
+            ):
+                return
+            current.cancelled = True
+        self._schedule(pointer, expire_time)
+
+    def _process_startup_queries(self) -> None:
+        if TYPE_CHECKING:
+            assert self._loop is not None
+        # This is a safety to ensure we stop sending queries if Zeroconf instance
+        # is stopped without the browser being cancelled
+        if self._browser.zc.done:
+            return
+
+        now_millis = current_time_millis()
+
+        # At first we will send 3 queries to get the cache populated
+        self._browser.async_send_ready_queries(True, now_millis, self._browser.types)
```

Comment:
Asking QU questions on the first go probably doesn't make sense since some devices simply won't respond to them